ads/feedad.js

```diff
@@ -0,0 +1,117 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the 'License');
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an 'AS-IS' BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {adConfig} from './_config';
+import {loadScript, validateData} from '../3p/3p';
+import {setStyle} from '../src/style';
+
+/**
+ * @typedef FeedAdGlobal
+ * @private
+ *
+ * @property {FeedAdAsync} feedad
+ */
+
+/**
+ * @typedef {Object} FeedAdAsync
+ * @private
+ *
+ * @property {FeedAd} [sdk]
+ * @property {!Function[]} cmd
+ */
+
+/**
+ * @typedef {Object} FeedAd
+ * @private
+ *
+ * @property {function(string)} init
+ * @property {function(string):Promise<FeedAdResponse>} requestAd
+ */
+
+/**
+ * @typedef {Object} FeedAdResponse
+ * @private
+ *
+ * @property {function():HTMLElement} createAdContainer()
+ */
+
+/**
+ * @typedef {Object} FeedAdData
+ * @private
+ *
+ * @property {string} clientToken
+ * @property {string} placementId
+ * @property {string} [background]
+ */
+
+/**
+ * @param {!FeedAdGlobal & Window} global
+ * @param {!FeedAdData} data
+ */
+export function feedad(global, data) {
+  validateData(data, ['clientToken', 'placementId'], ['background']);
+
+  global.feedad = global.feedad || {cmd: []};
+  global.feedad.cmd.push(() => {
+    global.feedad.sdk
+      .init(data.clientToken)
+      .then(() => global.feedad.sdk.requestAd(data.placementId))
+      .then((response) => {
+        const ad = response.createAdContainer();
+        const container = global.document.getElementById('c');
+        applyContainerStyle(container, data);
+        container.appendChild(ad);
+        global.context.renderStart();
```

Comment:
instead of doing `requestOptimalSize`, you could call `global.context.renderStart({width: xx, height: xx})` Rather than using `getBoundingRect`, you could pass in the size from your ad response (presumably you know the size of the creative?)

Reply:
Our ads are responsive, they only require a certain aspect ratio. This is why we would like to request the largest possible size instead of requesting the initial ad size.
I have now moved it into our SDK to a place where both the AMP size and the asset aspect ratio are known.